### PR TITLE
SOLR-15781: Document v2 API syntax conventions

### PR DESCRIPTION
**Pull request by gerlowskija**
https://issues.apache.org/jira/browse/SOLR-15781

# Description

Currently, decisions around what our v2 API should look like are scattered across a large number of JIRA tickets, PR reviews, and spreadsheets.  There's no one place giving an overview of the desired syntax (beyond perhaps the general guidance in SIP-16 to move towards "REST" where possible).

# Solution

This commit attempts to address this by introducing docs describing some of the conventions used in our v2 API.

Room for improvement remains: parameters conventions remain a big gap, as does sourcing to the original rationale behind many of these decisions.  But this PR provides a place to build off of moving forward.

# Tests

N/A - doc change only.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [x] I have added documentation to the dev-docs.

### CODE REVIEW

File: dev-docs/apis.adoc [hunk 1]
```diff
@@ -51,27 +51,28 @@ Separating the API "definition" and "implementation" in this way allows us to on
 
 Writing a new v2 API may appear daunting, but additions in reality are actually pretty simple:
 
-1. *Create POJO ("Plain Old Java Object") classes as needed to represent the API's request and response*:
+1. *Agree on Endpoint Syntax*: Before implementing, developers should generate consensus around what the new API will look like.  General guidelines and conventions for the v2 API syntax are described <<v2-api-conventions.adoc,here>>.
```
Comment (epugh) [inline dev-docs/apis.adoc:54]: In markdown you can do:

```
1. First item
1. Second item
1. Third item
1. Fourth item 
```

Does that work in adoc?  would that make life easier?

File: dev-docs/v2-api-conventions.adoc [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+Solr's "modify collection" functionality uses partial update semantics, but the v2 API uses `PUT` instead of `PATCH` due to the relative obscurity of the latter.
+
+For use within the v2 API, the four "popular" HTTP methods have the following semantics and implications:
+
+* `GET` - used for non-mutating (i.e. "read only") requests. Most often used to list elements of a particular resource type, or fetch information about about a specific named resource.
+* `POST` - used for non-idempotent resource modifications.
+* `PUT` - used for idempotent resource modifications.
+* `DELETE` - Used to delete or cleanup resource
+
+== Exceptional Cases - "Command" APIs
```
Comment (epugh) [inline dev-docs/v2-api-conventions.adoc:69]: Do we need to also call out the "index" special case, of being a collection or core in this doc?

**Review by epugh — CHANGES_REQUESTED** [review]
Two comments...

File: dev-docs/apis.adoc (hunk 1 above)
Comment (gerlowskija) [reply dev-docs/apis.adoc:54]: It definitely works in adoc.  The question I'd ask though is whether Github is smart enough to do it as well when it's previewing these asciidoc files.

That's the main way we expect these to be consumed, right?  Through the Github UI?  (i.e. we don't actually build these asciidoc files into any thing analogous to the ref-guide?)

I'll push the change and see how Github handles it; if GH doesn't do-the-right-thing I'll revert.

Comment (gerlowskija) [reply dev-docs/apis.adoc]: Looks like GH does do the right thing 👍

**Pull request merged**